modvendor/modvendor.go

```diff
@@ -0,0 +1,221 @@
+//
+// Vendored modules will be located at:
+//
+// - <vendordir>/<Source.Path>/<Source.Ref>
+//
+// If the provided source has no reference the provided Source.URL will be
+// used to retrieve the default remote branch to be used as reference.
+//
+// The whole path inside the vendor dir will be created if it not exists.
+// Vendoring is not recursive, so dependencies won't have their dependencies vendored.
+// Vendoring will also not download any git submodules.
+//
+// It returns the absolute path where the code has been vendored, which will be inside
+// the given vendordir.
+func Vendor(vendordir string, src Source) (string, error) {
+	logger := log.With().
+		Str("action", "modvendor.Vendor()").
+		Str("vendordir", vendordir).
+		Str("URL", src.URL).
+		Str("path", src.Path).
+		Str("ref", src.Ref).
+		Logger()
+
+	if src.Ref == "" {
+		// TODO(katcipis): handle default references.
+		// for now always explicit is fine.
+		return "", errors.E("src %v reference must be non-empty", src)
+	}
+
+	// TODO(katcipis): test that if vendor contains path with matching ref
+	// it will do nothing.
+
+	logger.Trace().Msg("setting up tmp workdir")
+
+	workdir, err := os.MkdirTemp("", "terramate-vendor")
+	if err != nil {
+		return "", errors.E(err, "creating workdir")
+	}
+	defer func() {
+		// We ignore the error here since after the final os.Rename
+		// the workdir will be moved and won't exist.
+		_ = os.Remove(workdir)
+	}()
+
+	clonedir := filepath.Join(vendordir, src.Path, src.Ref)
+
+	logger = logger.With().
+		Str("workdir", workdir).
+		Str("clonedir", clonedir).
+		Logger()
+
+	logger.Trace().Msg("setting up git wrapper")
+
+	g, err := git.WithConfig(git.Config{
+		WorkingDir: workdir,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	logger.Trace().Msg("cloning to workdir")
+
+	if err := g.Clone(src.URL, workdir); err != nil {
+		return "", err
+	}
+
+	if err := os.MkdirAll(filepath.Dir(clonedir), 0775); err != nil {
+		return "", errors.E(err, "creating mod dir inside vendor")
+	}
+
+	logger.Trace().Msg("moving cloned mod from workdir to clonedir")
+	if err := os.Rename(workdir, clonedir); err != nil {
+		// This may leave intermediary created dirs hanging on vendordir
+		// since we just create all and then delete clone dir on a failure to move.
+		// If we get a lot of errors from os.Rename we may need to handle this
+		// more gracefully, here we assume that os.Rename errors are rare since both
+		// dirs where just created.
+		errs := errors.L()
+		errs.Append(errors.E(err, "moving cloned module"))
+		errs.Append(os.Remove(clonedir))
+		return "", errs.AsError()
+	}
+	return clonedir, nil
+}
+
+// ParseSource parses the given modsource string.
+// The modsource must be a valid Terraform Git/Github source reference as documented in:
+//
+// - https://www.terraform.io/language/modules/sources
+//
+// Source references that are not Git/Github are not supported.
+func ParseSource(modsource string) (Source, error) {
```

Comment:
The idea is to use ParseSource in the future when we are parsing all module.source references from code and then we only vendor the ones we support.